### PR TITLE
Define consistent float ordering

### PR DESCRIPTION
**Pull request by robert3005**
We use total_cmp instead of partial_cmp to compare floating point numbers and use bit patern to determine float equality

### CODE REVIEW

File: vortex-array/src/array/primitive/stats.rs [hunk 1]
```diff
@@ -189,26 +190,27 @@ impl<T: PStatsType> StatsAccumulator<T> {
             self.nan_count += 1;
         }
 
-        if self.prev == next {
+        let prev_cmp = next.total_cmp(self.prev);
```
Comment (robert3005) [inline vortex-array/src/array/primitive/stats.rs]: In a followup need to introduce equals on nativeptype

File: vortex-scalar/src/lib.rs [hunk 2]
```diff
@@ -74,8 +74,11 @@ impl Scalar {
     }
 
     pub fn cast(&self, dtype: &DType) -> VortexResult<Self> {
-        if self.dtype() == dtype {
-            return Ok(self.clone());
+        if self.dtype().eq_ignore_nullability(dtype) {
```
Comment (robert3005) [inline vortex-scalar/src/lib.rs:81]: need to add test for this - changing nullability shouldn't touch the value

Comment (robert3005) [reply vortex-scalar/src/lib.rs:81]: this has to move after the next check

File: vortex-array/src/array/primitive/stats.rs [hunk 3]
```diff
@@ -189,26 +190,27 @@ impl<T: PStatsType> StatsAccumulator<T> {
...
+        } else if matches!(next.total_cmp(self.max), Ordering::Greater) {
             self.max = next;
         }
         self.prev = next;
     }
 
     pub fn finish(self) -> StatsSet {
         let is_constant = (self.min == self.max && self.null_count == 0 && self.nan_count == 0)
-            || self.null_count == self.len
-            || self.nan_count == self.len;
```
Comment (robert3005) [inline vortex-array/src/array/primitive/stats.rs:211]: we should never claim nan's can be made constant for now. While the comparison now uses bit patters which would mean that equal nans could be really equal the logic here has to change from pure is_nan comparison

**Comment by lwwmanning** [comment]
The downside though is that sorting an array of floats won't give what users probably expect?

**Comment by robert3005** [comment]
I don’t think anyone has expectation how nans sort. There’s no way to sort arbitrary float arrays without this

**Comment by lwwmanning** [comment]
Right, but the bit wise ordering of non-NaN floats isn't the same as the natural ordering?

**Comment by robert3005** [comment]
Hm, you're right. I think I went too far. You can define equality using bit equality

File: vortex-dtype/src/ptype.rs [hunk 4]
```diff
@@ -48,6 +48,10 @@ pub trait NativePType:
     const PTYPE: PType;
 
     fn is_nan(self) -> bool;
+
+    fn compare(self, other: Self) -> Ordering;
+
+    fn is_eq(self, other: Self) -> bool;
```
Comment (AdamGS) [inline vortex-dtype/src/ptype.rs:54]: Can't these two be expressed with additional trait bounds?

Comment (robert3005) [reply vortex-dtype/src/ptype.rs:54]: We can define an additional traits but afaik there isn't one that has the behaviour as implemented here

Comment (AdamGS) [reply vortex-dtype/src/ptype.rs:54]: Isn't this just `Ord`?

Comment (AdamGS) [reply vortex-dtype/src/ptype.rs:54]: ok I got it, its about the custom implementation

**Review by AdamGS — APPROVED** [review]

File: vortex-dtype/src/ptype.rs [hunk 5]
```diff
@@ -30,7 +31,6 @@ pub enum PType {
 pub trait NativePType:
     Send
     + Sync
-    + Sized
```
Comment (lwwmanning) [inline vortex-dtype/src/ptype.rs:33]: Why no longer `Sized`?

Comment (robert3005) [reply vortex-dtype/src/ptype.rs:33]: Because it's always Sized https://doc.rust-lang.org/std/marker/trait.Sized.html

**Pull request merged**